### DISCUSSION TITLE
手工编译 quick-sort 到 python

### DISCUSSION THREAD

**Discussion by nobodxbodon** [yuyan-lang/yuyan#76] [Q&A] [answered]
尝试按照语义手工将 `performance-investigation/quick-sort/quick-sort.yuyan` 的主要部分转译到 python ：
```
快排乃化整数列而整数列也。
快排者会甲而
    鉴甲而
        有【】则【】
        或有头衔尾则
            （滤于（会数而数小于头）于尾）
            接（滤于（会数而数等于头）于尾）
            接（滤于（会数而数大于头）于尾）也。
```

转译为（这里“尾”的语义是不同于yuyan代码吗？否则有bug，头元素缺失）：
```
def 快排(arr: list[int]) -> list[int]:
    match arr:
        case []:
            return []
        case [头, *尾]:
            更小 = list(filter(lambda x: x < 头, 尾))
            相等 = list(filter(lambda x: x == 头, 尾))
            更大 = list(filter(lambda x: x > 头, 尾))
            return 更小 + 相等 + 更大
```
将此段替换 `performance-investigation/quick-sort/quick-sort.py` 的 `quicksort(arr)` 后，运行效率比它还高十倍的样子（py3.12.1)，参数从 1000 到 10000000。

另外，强烈建议尽早整理报错机制和改进反馈。本想如下加个打印输出看快排后数列内容，结果报错超长完全看不清：
```
结果的定义是快排于「数列」。
...
打印行于结果。
```

**Comment by nobodxbodon** [comment]
把函数最后一行改为了如下，否则没递归差太大：
```python
return 快排(更小) + [头] + 相等 + 快排(更大)
```

测试 1000000 结果 10 秒，原 `performance-investigation/quick-sort/quick-sort.py` 三秒。这样合理些。

**Comment by UltimatePea** [comment, marked as answer]
> 转译为（这里“尾”的语义是不同于yuyan代码吗？否则有bug，头元素缺失）：

确实有bug，应该把头加进去

> 另外，强烈建议尽早整理报错机制和改进反馈。本想如下加个打印输出看快排后数列内容，结果报错超长完全看不清：

是的，我们有很多开发工作都因为运行速度缓慢而延后了，报错的话只要看前几行加中间几行就行，开发阶段报错机制需要详细一点因为通常你搞不清楚是编译器出错了还是程序写错了

> 本想如下加个打印输出看快排后数列内容，结果报错超长完全看不清

打印行只能打印字符串，我们还没做interface之类的功能，所以要把数列转换成字符串之后才能打印
